File: ruby/missing_ranges.rb
# Given a sorted integer array where the range of elements are [lower, upper]
# inclusive, return its missing ranges.

# For example, given [0, 1, 3, 50, 75], lower = 0 and upper = 99, return ["2",
# "4->49", "51->74", "76->99"].
#
# Idea:
# Add two "artificial" elements, lower -1 before the first element and upper + 1
# after the last element

def missing_ranges(array, lower = 0, upper = 99)
  ranges = []

  i = 0
  prev = lower - 1

  while i <= array.size
    curr = (i == array.size) ? upper + 1 : array[i]

    if curr - prev > 1
      ranges << get_range(prev + 1, curr - 1)
    end

    prev = curr
    i += 1
  end

  ranges
end

def get_range(from, to)
  from == to ? "#{from}" : "#{from}->#{to}"
end

a = [0, 1, 3, 50, 75]
puts missing_ranges(a)

File: java/MissingRanges.java
/*
Note:
1. in the question, return value is a list of strings
2. If range is [start, end], add two "artificial" elements, start - 1, end + 1

 */

import java.util.List;
import java.util.ArrayList;

public class MissingRanges {
  public List<String> findMissingRanges(int[] array, int startRange, int endRange) {
    List<String> ranges = new ArrayList<String>();

    int prev = startRange - 1;
    for (int i = 0; i <= array.length; i++) {
      int current = (i == array.length) ? endRange + 1 : array[i];

      if (current - prev > 1) {
        ranges.add(getRange(prev + 1, current - 1));
      }
      prev = current;
    }

    return ranges;
  }

  private String getRange(int from, int to) {
    return (from == to) ? String.valueOf(from) : String.format("%d->%d", from, to);
  }

  public static void main(String[] args) {
    MissingRanges finder = new MissingRanges();
    int[] test1 = {0, 1, 3, 50, 75};

    List<String> ranges = finder.findMissingRanges(test1, 0, 100);
    for(String range : ranges) {
      System.out.println(range);
    }

    ranges = finder.findMissingRanges(new int[0], 0, 100);
    for(String range : ranges) {
      System.out.println(range);
    }
  }
}
